thunderbirdparser/src/org/sleuthkit/autopsy/thunderbirdparser/EmailMessageThreader.java

```diff
@@ -0,0 +1,674 @@
+                    if (childMessage != null) {
+                        subject = childMessage.getSimplifiedSubject();
+                        if (!subject.isEmpty()) {
+                            reSubject = childMessage.isReplySubject();
+                            break;
+                        }
+                    }
+                }
+            }
+
+            if (subject == null || subject.isEmpty()) {
+                continue;  // Give up on this container
+            }
+
+            Container tableContainer = subject_table.get(subject);
+
+            if (tableContainer == null
+                    || // Not in table
+                    (tableContainer.getMessage() != null && rootSetContainer.getMessage() == null)
+                    || // One in the table is empty, but current is not
+                    (!reSubject && (tableContainer.getMessage() != null && tableContainer.getMessage().isReplySubject()))) { //current doesn't have RE in it, use current instead of the one in the table
+                subject_table.put(subject, rootSetContainer);
+            }
+
+        }
+
+        return subject_table;
+    }
+    
+    private void assignThreadIDs(Set<Container> containerSet, String IDPrefix) {
+        int threadCounter = 0;
+        
+        for(Container container: containerSet) {
+            String threadID = String.format("%s-%d", IDPrefix, threadCounter++);
+            addThreadID(container, threadID);
+        }
+    }
+    
+    private void addThreadID(Container container, String threadID) {
+        if(container == null) {
+            return;
+        }
+        
+        EmailMessage message = container.getMessage();
+        if(message != null) {
+            message.setMessageThreadID(threadID);
+        }
+        
+        if(container.hasChildren()) {
+            for(Container child: container.getChildren()) {
+                addThreadID(child, threadID);
+            }
+        }
+    }
+
+    /**
+     * Prints a set of containers and their children.
+     *
+     * @param containerSet Set of containers to print out
+     * @param prefix       A prefix for each line to show child depth.
+     */
+    private void printContainerSet(Set<Container> containerSet, String prefix) {
+        containerSet.stream().map((container) -> {
+            if (container.getMessage() != null) {
+                System.out.println(prefix + container.getMessage().getSubject());
+            } else {
+                System.out.println("<Empty Container>");
+            }
+            return container;
+        }).filter((container) -> (container.hasChildren())).forEachOrdered((container) -> {
+            printContainerSet(container.getChildren(), prefix + "\t");
+        });
+    }
+
+    /**
+     * The container object is used to wrap and email message and track the
+     * messages parent and child messages.
+     */
+    final class Container {
```

Comment:
I think there should be "Email" somewhere in the name of this class to make it more clear it represents a single Email.

Reply:
I took a lot of the naming conventions straight from the algorithm docs so as to not confuse myself\remember I named it something different.  

If you would like Email in the name I can change it.